src/python/grpcio/grpc/_common.py

```diff
@@ -76,9 +76,41 @@
 }
 
 
-def metadata(application_metadata):
+def encode(s):
+  if isinstance(s, bytes):
+    return s
+  else:
+    return s.encode('ascii')
+
+
+def decode(b):
+  if isinstance(b, str):
+    return b
+  else:
+    try:
+      return b.decode('utf8')
+    except UnicodeDecodeError:
+      logging.exception('Invalid encoding on {}'.format(b))
+      return b.decode('latin1')
+
+
+def cygrpc_metadata(application_metadata):
   return _EMPTY_METADATA if application_metadata is None else cygrpc.Metadata(
-      cygrpc.Metadatum(key, value) for key, value in application_metadata)
+      cygrpc.Metadatum(encode(key), encode(value))
+          for key, value in application_metadata)
+
+
+def application_metadata(cygrpc_metadata):
+  if cygrpc_metadata is None:
+    return []
+  else:
+    result = []
+    for key, value in cygrpc_metadata:
+      if key[-4:] == b'-bin':
+        result.append((decode(key), value))
+      else:
+        result.append((decode(key), decode(value)))
+    return result
```

Comment:
How about `return tuple(decoded_metadata)`? Let's avoid returning a mutable value unless we're aware of some use case that really warrants it?